### PR TITLE
feat(chat): unify LLM credential resolution into one chokepoint

### LINKED ISSUE #221: Custom providers with `parent_provider` don't inherit the parent's SDK, causing 404 on Anthropic-style endpoints
### Summary

When a user defines a `custom_providers` entry with `parent_provider: "anthropic"` (or any non-OpenAI SDK parent) and routes a custom model through it, chat requests fail with HTTP 404 and the error payload shows `provider_module: "openai"`.

The root cause is that `from_custom_config` (`src/llms/llm.py`) reads the SDK from `providers.json` keyed by `config["provider"]`. User-defined custom provider slugs are not present in the manifest, so `get_provider_info()` returns `{}` and `sdk` falls back to the default `"openai"`. The OpenAI client then POSTs to `/chat/completions` against an endpoint that only speaks Anthropic protocol (`/v1/messages`), returning 404.

### Reproduction

1. Configure a custom provider in Settings:

   ```json
   {
     "custom_providers": [
       {"name": "anthropic-eval-provider", "parent_provider": "anthropic"}
     ],
     "custom_models": [
       {"name": "eval-deepseek-v4-pro", "model_id": "deepseek-v4-pro", "provider": "anthropic-eval-provider"}
     ]
   }
   ```

2. Store a BYOK key + Anthropic-style `base_url` for `anthropic-eval-provider` via `PUT /api/v1/users/me/api-keys`.

3. Trigger a chat request:

   ```bash
   curl -N "http://localhost:8000/api/v1/threads/messages" \
     -H "Content-Type: application/json" \
     -d '{
       "workspace_id": "<workspace>",
       "agent_mode": "ptc",
       "llm_model": "eval-deepseek-v4-pro",
       "messages": [{"role": "user", "content": "hi"}]
     }'
   ```

### Expected

Request is sent to `<base_url>/v1/messages` using the Anthropic SDK and returns the model reply.

### Actual

```
Error code: 404 - {'status': 404, 'error': 'Not Found', 'path': '/anthropic/v1/chat/completions'}
provider_module: openai
```

The OpenAI SDK targets `/chat/completions`, which the upstream endpoint does not serve.

### Root cause

`src/server/handlers/chat/llm_config.py::_resolve_custom_model_byok` has three BYOK key-lookup paths:

- **Path 1**: the model name itself is a custom provider slug
- **Path 2**: `custom_models.provider` is a custom provider slug (this bug's trigger)
- **Path 3**: system provider sibling fan-out

Path 3 already rewrites the provider after finding a key:

```python
if candidate != provider:
    custom_config = {**custom_config, "provider": candidate}
```

so `from_custom_config` reads SDK / default_headers / `use_response_api` from the manifest entry that actually holds the key. Paths 1 and 2 omit this step. For an Anthropic-parented custom provider, `custom_config["provider"]` remains the user's slug, so `sdk` defaults to `"openai"`.

### Impact

Any BYOK setup that proxies an Anthropic-style endpoint through a user-defined custom provider (`parent_provider: "anthropic"` or any other non-OpenAI SDK parent). In the current manifest this means all internal / corporate Anthropic-compatible gateways.

### PR DESCRIPTION
**Pull request by Chen-zexi**
## Summary

A BYOK/OAuth user's own key was silently replaced by the platform key on several
LLM paths. These weren't independent bugs — they were divergences from an
architecture that already half-existed. This finishes it: one resolution
primitive, one role accessor, roles-as-data, and a CI guard so new code can't
reintroduce a divergent path.

**Core abstraction**
- `CredentialSource` (OAUTH / BYOK / PLATFORM / NONE) — first-class answer to
  "did the user's credential pay?", distinct from `ModelSource` (which classifies
  the model). Stamped on `AgentConfig.credential_source` by the resolver.
- `resolve_model_client()` — single primitive returning `ResolvedClient(client,
  model_source, credential_source)`. One OAuth → BYOK → platform ladder replaces
  three duplicated platform-fallback spots.
- `role_registry()` + `LLMRole` — compaction, fetch, and `subagent:<name>` resolve
  through one loop with one BYOK-pure fallback policy. Subagents stop being a
  special case.
- `client_for_role()` accessor on `AgentConfig` — consumers read role clients
  through one method; the BYOK-pure fallback is materialized once at write time.

**Bug fixes that fall out of it**
- **Credit-gate bypass (`threads.py`):** gate now keys off `credential_source`,
  not `llm_client is not None`. A non-BYOK reasoning user built a platform client,
  which made `llm_client != None` → wrongly treated as BYOK → hit the BYOK
  negative-balance gate instead of the platform daily-credit gate. Closed.
- **Automation (`automation_executor.py`):** previously skipped the credit gate
  entirely and never attempted BYOK. Now computes `has_cred = is_byok_active OR
  has_any_oauth_token`, gates with `enforce_credit_limit` before the workflow, and
  passes `is_byok=has_cred` to both workflow calls. OAuth-only users no longer
  mis-gated as platform.
- **Coding-variant BYOK (`resolve_byok_llm_client`):** system branch now walks
  [provider → parent → sibling] to find the key, but pins SDK/base_url to the
  model's own provider. Fixes silent platform fallback for `*-coding` models whose
  key lives under the variant slug.
- **Custom-provider SDK (#221):** custom providers inherit their manifest parent's
  SDK so an Anthropic-parented gateway no longer builds an OpenAI client pointed at
  an Anthropic endpoint (404). OpenAI-compatible gateways are left untouched (no
  forced Responses API).
- **Fetch / compaction:** both now use the materialized role client for BYOK users
  (their key pays) and the cheap name-based platform model for system users.
- **Insight fallback:** routed through `LLMService.complete` with a threaded
  `user_id` — deletes the last raw `create_llm` in the insight path.

**Forever guard**
- `test_llm_constructor_boundary.py` — AST scan of `ptc_agent/**`, `tools/**`,
  `server/services/**` for request-time LLM constructors off an allowlist. New
  divergent paths fail CI.

## Diff Size
- Total: 31 files changed, 3038 insertions(+), 416 deletions(-)
- Net (excl. tests): 14 files changed, 749 insertions(+), 393 deletions(-)

## Test Coverage
3555 backend unit tests pass. 15 test files added/changed covering:
- `resolve_model_client` (OAuth-first / BYOK-second / platform-only-when-allowed)
- `credential_source` matrix (OAuth, BYOK, BYOK-on-system-model, non-BYOK+reasoning
  → PLATFORM, non-BYOK no-reasoning → NONE)
- BYOK-pure materialization gated on `credential_source` (PLATFORM client NOT
  copied into roles — regression guard)
- coding-variant walk + SDK-trap guard (neutral placeholder providers)
- `threads.py` credit-gate regression (platform daily vs BYOK negative-balance)
- automation gate (BYOK-only / OAuth-only / neither + gate-before-workflow)
- subagent instance injection reaches `create_agent` (not the string)
- fetch context-var contract + copy-before-mutate
- `_byok_cache` miss safety (custom provider resolves via direct lookup)
- the constructor-boundary AST guard + self-test

## Live verification (against a running backend, BYOK user)
- Web fetch: extraction sub-call billed `billing_type: byok`, 0 platform cost
- Subagent: research subagent's calls billed byok, 0 platform cost
- Automation (happy path): gate passed, run billed byok, `total_credits=0`
- Automation (gate block): forced negative balance → execution `failed` with
  `429 negative_balance`, no usage row created, workflow never ran
- Platform billing (prod manifest): `is_byok=false`, credits charged

## Out of scope (tracked, deferred)
- Platform-credit deduction (cross-service ginlix-platform integration) — this
  change only gates and emits the `metered_to_user` signal; no code deducts yet.
- Thread-less one-shot / insight usage recording — needs a thread-less usage path
  AND the absent deduction consumer; designed together later.
- OpenAI O1/O3 Responses-API edge cases, DELETE-key cache invalidation, HITL resume
  BYOK threading — documented in the plan, not triggered by current models.

## Test plan
- [x] 3555 backend unit tests pass (`PYTHONPATH=src pytest tests/unit/ -q`)
- [x] Live e2e: BYOK billing on main model, fetch, subagent, automation
- [x] Live e2e: automation credit gate blocks (no free run)
- [x] Live e2e: platform billing charges credits (prod manifest)

## Summary by CodeRabbit

* **New Features**
  * Per‑subagent model selection so agents can use role-specific models.

* **Bug Fixes**
  * Fixed credential-source handling and BYOK key discovery across provider variants.
  * Clearer user-facing errors when a selected model is unavailable, with a Settings CTA.

* **Infrastructure**
  * Improved LLM credential resolution, compaction client selection, and credit‑gating for automation workflows.

### CODE REVIEW

File: src/server/handlers/chat/model_availability.py [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+        # Residual race window: between the re-read above and this upsert, a
+        # Settings save could still land and get overwritten by our ``None``
+        # delete. Narrow (single DB read → single DB write) and self-healing
+        # (the user saves again and it sticks). Not worth a CTE or advisory
+        # lock for the size of the hole.
+        await upsert_user_preferences(user_id=user_id, other_preference=updates)
+        await invalidate_user_prefs_cache(user_id)
+        logger.info(
+            f"[CHAT] Scrubbed stale model prefs for user={user_id}: {removed}"
+        )
```
[the remark this answers is not in the document]
Comment (Chen-zexi) [reply src/server/handlers/chat/model_availability.py:112]: Declined. `user_id` here is an internal UUID already logged at info level throughout this codebase (`onboarding.py`, `user_data_io.py`, `workspace_manager.py`), and `removed` is a list of model-preference *names* (e.g. `gpt-4o`), not PII. Reducing it to a count would drop the debugging signal (which prefs were scrubbed) without removing a real exposure. No change — this matches the established logging convention.

File: tests/unit/test_llm_constructor_boundary.py [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+
+# Repo root = two levels up from this file (tests/unit/).
+REPO_ROOT = os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
+
+# Dirs scanned for request-time constructor calls (relative to repo root).
+SCAN_ROOTS = (
+    "src/ptc_agent",
+    "src/tools",
+    "src/server/services",
+)
```
[the remark this answers is not in the document]
Comment (Chen-zexi) [reply tests/unit/test_llm_constructor_boundary.py:23]: Partially applied in 0623397d. Added the fail-loud guard for a missing scan root (a renamed/moved root now appends to `parse_errors`, which `test_no_unallowlisted_llm_constructors` asserts is empty — so coverage can no longer be silently dropped).

Declined the `server/app` + `server/handlers` expansion: `server/handlers/chat/` is the credential-*aware* resolver layer — it is *supposed* to call `create_llm` / build the platform fallback. The guard's purpose is to keep request-time constructors out of credential-*blind* code (`ptc_agent`, `tools`) and `server/services`. Scanning the resolver would require allowlisting the entire resolution engine, which defeats the guard.

**Comment by Chen-zexi** [comment]
Verdicts on the review findings:

**Blocking — sequential role/fallback resolution** (fixed in `0623397d`): Confirmed regression — `origin/main` resolved all subsidiary + fallback models in one `asyncio.gather`; the role-registry refactor turned it into sequential `await` loops. Restored `gather` in both `_resolve_role_clients` and `_resolve_fallback_clients`, and parallelized the two top-level calls (they write disjoint config fields). Also fed subagent model slugs into the BYOK batch prefetch (role list built once, shared by prefetch + resolver).

**Non-blocking #2 — `model_copy()` relies on undocumented invariant** (fixed in `2e28ff62`): Added an explicit `config.llm_client is not None` guard alongside the OAUTH/BYOK `credential_source` check — the documented invariant is now an enforced guard.

**Non-blocking #3 — OAuth-only users waste a BYOK prefetch** (fixed in `2e28ff62`): Workflow `is_byok` now keys off `has_byok`; the credit gate keeps `has_cred` (OAuth users stay credentialed). Test updated to assert the split (gate `byok=True`, workflow `is_byok=False`).

**Non-blocking #1 — transient DB error in cred check counts as automation failure**: Deferred. Valid, but the entire `try` block (workspace resolve, workflow drain, persistence wait) shares the same infra-error→failure exposure, and the 429 gate-rejection is *intentionally* counted (see `automation_executor.py:262`). Spot-patching only the cred-check `gather` would be inconsistent; this belongs to a broader transient-error-classification pass rather than this PR.

All 3555 unit tests pass; `ruff` clean.
